In [171]:
# import libraries
import pandas as pd
import numpy as np
import folium
import geopy
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
import requests # library to handle requests



Get the data from Wikipedia with pandas


In [172]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
list_df = pd.read_html(url)

The created list of Data Frames contain 3 data frames. The first one is the one we need. Let's assign it to a new data frame.



In [173]:
df = list_df[0]
print(df.shape)
# Display the first 10 rows
df.head()

(180, 3)


Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Let's rename the first column to PostalCode



In [174]:
df.rename(columns={"Postal Code":"PostalCode"},inplace=True)

# Get rid of rows with no Borough
to_drop = np.where(df['Borough']=='Not assigned')[0]
df.drop(to_drop,axis=0,inplace=True)
df.shape

(103, 3)

Let's check if any "Not assigned" values are still in Borough...

In [175]:
(df['Borough']=='Not assigned').sum()

0

Let's reset the index and check how the new DF looks like...

In [176]:
df = df.reset_index(drop=True)
df.head()

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Let's check now for "Not assigned" values in Neighbourhood

In [177]:
(df['Neighbourhood']=="Not assigned").sum()

0

Now we combine rows that have the same PostalCode value

In [178]:
df = df.groupby(['PostalCode','Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.head()

PostalCode      Borough                           Neighbourhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [179]:
df.shape

(103, 3)

# ==== End of question 1 ====

# ==== Question 2 =====

 The geocoder package didn't work so we will load the csv data directly

In [180]:
coord_data = pd.read_csv('Geospatial_Coordinates.csv')
coord_data.rename(columns={"Postal Code": "PostalCode" },inplace=True)

In [181]:
coord_data.head(10)

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476
5        M1J  43.744734 -79.239476
6        M1K  43.727929 -79.262029
7        M1L  43.711112 -79.284577
8        M1M  43.716316 -79.239476
9        M1N  43.692657 -79.264848

Now we merge both data frames into a single data frame....

In [182]:
new_df = df.merge(coord_data,on='PostalCode')
new_df.head(10)

PostalCode      Borough                                    Neighbourhood  \
0        M1B  Scarborough                                   Malvern, Rouge   
1        M1C  Scarborough           Rouge Hill, Port Union, Highland Creek   
2        M1E  Scarborough                Guildwood, Morningside, West Hill   
3        M1G  Scarborough                                           Woburn   
4        M1H  Scarborough                                        Cedarbrae   
5        M1J  Scarborough                              Scarborough Village   
6        M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park   
7        M1L  Scarborough                  Golden Mile, Clairlea, Oakridge   
8        M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West   
9        M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

# ==== End of question 2 ====

# ==== Question 3 ====

### First, we select the Boroughs that have the word "Toronto" and create a new data frame to work with...

In [221]:
neighborhoods = new_df[new_df['Borough'].str.contains("Toronto")].reset_index(drop=True)
print(neighborhoods.shape)
neighborhoods.head(10)

(39, 5)


PostalCode          Borough  \
0        M4E     East Toronto   
1        M4K     East Toronto   
2        M4L     East Toronto   
3        M4M     East Toronto   
4        M4N  Central Toronto   
5        M4P  Central Toronto   
6        M4R  Central Toronto   
7        M4S  Central Toronto   
8        M4T  Central Toronto   
9        M4V  Central Toronto   

                                       Neighbourhood   Latitude  Longitude  
0                                        The Beaches  43.676357 -79.293031  
1                       The Danforth West, Riverdale  43.679557 -79.352188  
2                     India Bazaar, The Beaches West  43.668999 -79.315572  
3                                    Studio District  43.659526 -79.340923  
4                                      Lawrence Park  43.728020 -79.388790  
5                                   Davisville North  43.712751 -79.390197  
6                  North Toronto West, Lawrence Park  43.715383 -79.405678  
7                                         Davisville  43.704324 -79.388790  
8                        Moore Park, Summerhill East  43.689574 -79.383160  
9  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049

There are 39 Neighbourhoods that we will be working with (note that most Neighbourhoods are a collection of many smaller neighbourhoods)

### Now we use geopy and get the coordinates for Toronto

In [184]:
from geopy.geocoders import Nominatim

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Use Folium to mark the 39 different neighbourhoods...

In [185]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Now we use the Foursquare IPA, Let's define Foursquare Credentials and Version

In [226]:
import foursquare_keys  # python script that contains my personal CLIENT IDs
CLIENT_ID = foursquare_keys.CLIENT_ID # your Foursquare ID
CLIENT_SECRET = foursquare_keys.CLIENT_SECRET # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

#### Let's create a function to get the nearby Venues to all the neighborhoods in Toronto



In [227]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print("Venues found for",name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Apply the function to our Data Frame

In [229]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighbourhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Venues found for The Beaches
Venues found for The Danforth West, Riverdale
Venues found for India Bazaar, The Beaches West
Venues found for Studio District
Venues found for Lawrence Park
Venues found for Davisville North
Venues found for North Toronto West, Lawrence Park
Venues found for Davisville
Venues found for Moore Park, Summerhill East
Venues found for Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Venues found for Rosedale
Venues found for St. James Town, Cabbagetown
Venues found for Church and Wellesley
Venues found for Regent Park, Harbourfront
Venues found for Garden District, Ryerson
Venues found for St. James Town
Venues found for Berczy Park
Venues found for Central Bay Street
Venues found for Richmond, Adelaide, King
Venues found for Harbourfront East, Union Station, Toronto Islands
Venues found for Toronto Dominion Centre, Design Exchange
Venues found for Commerce Court, Victoria Hotel
Venues found for Roselawn
Venues found for Forest Hill North & Wes

Let's see how many total Venues were found...

In [224]:
print(toronto_venues.shape)
toronto_venues.head()

(1624, 7)


Neighbourhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

The total number of Venues is 1624 ... Now let's see how many Venues were returned for each neighbourhood

In [191]:
toronto_venues.groupby('Neighbourhood').count()[['Venue']]

Venue
Neighbourhood                                            
Berczy Park                                            55
Brockton, Parkdale Village, Exhibition Place           23
Business reply mail Processing Centre, South Ce...     16
CN Tower, King and Spadina, Railway Lands, Harb...     16
Central Bay Street                                     68
Christie                                               16
Church and Wellesley                                   75
Commerce Court, Victoria Hotel                        100
Davisville                                             33
Davisville North                                        9
Dufferin, Dovercourt Village                           13
First Canadian Place, Underground city                100
Forest Hill North & West, Forest Hill Road Park         4
Garden District, Ryerson                              100
Harbourfront East, Union Station, Toronto Islands     100
High Park, The Junction South                          25
India Bazaar, The Beaches West                         19
Kensington Market, Chinatown, Grange Park              74
Lawrence Park                                           3
Little Portugal, Trinity                               45
Moore Park, Summerhill East                             2
North Toronto West, Lawrence Park                      18
Parkdale, Roncesvalles                                 14
Queen's Park, Ontario Provincial Government            33
Regent Park, Harbourfront                              44
Richmond, Adelaide, King                              100
Rosedale                                                4
Roselawn                                                2
Runnymede, Swansea                                     33
St. James Town                                         85
St. James Town, Cabbagetown                            48
Stn A PO Boxes                                         96
Studio District                                        37
Summerhill West, Rathnelly, South Hill, Forest ...     14
The Annex, North Midtown, Yorkville                    19
The Beaches                                             4
The Danforth West, Riverdale                           43
Toronto Dominion Centre, Design Exchange              100
University of Toronto, Harbord                         34

In [192]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


Let's see which categories are the most common, and which are the least common...

In [252]:
pd.DataFrame({'Counts':pd.Series(toronto_venues['Venue Category'].value_counts())}).sort_values(by='Counts',ascending=False)[:10]

Counts
Coffee Shop             159
Café                     84
Restaurant               56
Hotel                    41
Italian Restaurant       35
Park                     35
Japanese Restaurant      32
Bakery                   31
Bar                      28
Pizza Place              27

The most common Venues are coffee shops, Italian/Japanese rastaurants, Parks, Bakeries, Bars and hotels. 

In [253]:
pd.DataFrame({'Counts':pd.Series(toronto_venues['Venue Category'].value_counts())}).sort_values(by='Counts',ascending=True)[:10]

Counts
Filipino Restaurant       1
Theme Restaurant          1
Bus Line                  1
Malay Restaurant          1
College Cafeteria         1
Gas Station               1
Massage Studio            1
Sake Bar                  1
Opera House               1
College Auditorium        1

The least common Venues are Exotic food restaurants, Bus line, gas station, and other specific uncommon venues...

### Let's analyze each neighborhood with one hot enconding



In [193]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(1624, 238)


Neighbourhood  Afghan Restaurant  Airport  \
0                   The Beaches                  0        0   
1                   The Beaches                  0        0   
2                   The Beaches                  0        0   
3                   The Beaches                  0        0   
4  The Danforth West, Riverdale                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Theme Restaurant  \
0                 0                    0             0  ...                 0   
1                 0                    0             0  ...                 0   
2                 0                    0             0  ...                 0   
3                 0                    0             0  ...                 0   
4                 0                    0             0  ...                 0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      1              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  \
0                 0                      0         0              0   
1                 0                      0         0              0   
2                 0                      0         0              0   
3                 0                      0         0              0   
4                 0                      0         0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 238 columns]

### Now we group the data by Neighbourhoods and create the Data Frame we will use for the clustering function. 

In [194]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013333   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0625              0.0625        0.0625           0.125   
4    0.0000              0.0000        0.0000           0.000   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000             

In [195]:
toronto_grouped.shape

(39, 238)

As expected, the DataFrame has 39 rows (neighbourhoods) and 238 columns (for each unique venue category)

### Let's see which are the 5 most common Venue categories in each Neighbourhood...

In [196]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1    Cocktail Bar  0.04
2        Beer Bar  0.04
3  Farmers Market  0.04
4     Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3       Nightclub  0.09
4       Pet Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0  Gym / Fitness Center  0.06
1         Auto Workshop  0.06
2           Pizza Place  0.06
3            Comic Shop  0.06
4      Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0       Airport Lounge  0.12
1      Airport Service  0.12
2  Rental Car Location  0.06
3        Boat or Ferry  0.06
4   Airport Food Court  0.06


----Central Bay Street----
                venue  freq
0         Coff

### Now we create a new DataFrame that contains the 10 most common Venue categories for each Neighbourhood...

In [216]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...  Gym / Fitness Center   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant                Bakery   
1             Nightclub           Coffee Shop        Breakfast Spot   
2         Auto Workshop            Comic Shop                  Park   
3       Airport Service              Boutique                 Plane   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar        Farmers Market            Restaurant   
1         Grocery Store          Intersection                   Bar   
2           Pizza Place      Recording Studio            Restaurant   
3               Airport    Airport Food Court          Airport Gate   
4           Salad Place      Department Store       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop    Basketball Stadium    Sporting Goods Shop  
1                Bakery    Italian Restaurant           Climbing Gym  
2               Butcher         Burrito Place                Brewery  
3      Airport Terminal                   Bar        Harbor / Marina  
4          Burger Joint       Bubble Tea Shop    Japanese Restaurant

### Finally, we use sklearn and run the K-means cluster function to the Neighborhoods...
We will use 4 clusters because a larger number results in many clusters with only 1 member, which may be an overfit model.


In [217]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check the distribution of cluster labels generated 
pd.Series(kmeans.labels_[0:40]).value_counts()

1    34
3     3
2     1
0     1
dtype: int64

As seen from the distribution above, most neighbourhoods (34/39) have been clustered in the same group. The other clusters contain few neighbourhoods.

### Now let's add the latitude, longitude, and Cluster label information to the DataFrame created above, so we can feed it to Folium and mark the clusters.

In [218]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() 

PostalCode          Borough                   Neighbourhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                   Pub     Health Food Store   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1                  Park      Sushi Restaurant   
3 -79.340923               1           Coffee Shop   American Restaurant   
4 -79.388790               1                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Neighborhood           Yoga Studio   
1    Italian Restaurant            Restaurant        Ice Cream Shop   
2     Fish & Chips Shop            Steakhouse                   Pub   
3                Bakery               Brewery                  Café   
4           Swim School    Dim Sum Restaurant  Ethiopian Restaurant   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0     Distribution Center    Dim Sum Restaurant                        Diner   
1  Furniture / Home Store             Bookstore                    Juice Bar   
2                 Brewery  Fast Food Restaurant           Italian Restaurant   
3               Gastropub           Yoga Studio                  Fish Market   
4             Escape Room     Electronics Store  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store       Doner Restaurant  
1                   Spa                    Pub  
2            Restaurant            Pizza Place  
3                  Park           Neighborhood  
4   Dumpling Restaurant             Donut Shop

### Let's display the map and mark the different clusters with different colors.

In [255]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
rainbow = ['red','blue','orange','green']


# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters